### PR TITLE
New command Register Scheme

### LINKED ISSUE #18298: `Search is:Scheme` lists inductives not schemes
~~~coq
Search is:Scheme.
(* 
Number.number: Set
Number.signed_int: Set
Number.uint: Set
False: Prop
True: Prop
Hexadecimal.hexadecimal: Set
Decimal.decimal: Set
Decimal.signed_int: Set
Hexadecimal.signed_int: Set
Hexadecimal.uint: Set
nat: Set
bool: Set
Decimal.uint: Set
comparison: Set
unit: Set
Empty_set: Set
Byte.byte: Set
list: Type -> Type
option: Type -> Type
inhabited: Type -> Prop
or: Prop -> Prop -> Prop
and: Prop -> Prop -> Prop
sumor: Type -> Prop -> Type
sumbool: Prop -> Prop -> Set
prod: Type -> Type -> Type
sum: Type -> Type -> Type
eq_true: bool -> Prop
sig: forall [A : Type], (A -> Prop) -> Type
sigT: forall [A : Type], (A -> Type) -> Type
ex: forall [A : Type], (A -> Prop) -> Prop
eq: forall {A : Type}, A -> A -> Prop
le: nat -> nat -> Prop
BoolSpec: Prop -> Prop -> bool -> Prop
sigT2: forall [A : Type], (A -> Type) -> (A -> Type) -> Type
ex2: forall [A : Type], (A -> Prop) -> (A -> Prop) -> Prop
Acc: forall [A : Type], (A -> A -> Prop) -> A -> Prop
sig2: forall [A : Type], (A -> Prop) -> (A -> Prop) -> Type
CompareSpecT: Prop -> Prop -> Prop -> comparison -> Set
CompareSpec: Prop -> Prop -> Prop -> comparison -> Prop
*)
~~~

### PR DESCRIPTION
**Pull request by SkySkimmer**
This allows to register constants in the scheme mechanism, such that the scheme generator does not run.

Useful to work around bugs like https://github.com/coq/coq/issues/5341: instead of having to write a dummy definition to get subst to generate its schemes we can define them directly, and use alternate definitions if we want.

See also https://github.com/HoTT/Coq-HoTT/pull/1794

The scheme kinds are currently low level and undocumented. This is IMO acceptable in the same way as we accept undocumented names in the regular `Register`.
It would be nice to stabilize and document some (especially for induction principles, then we could deprecate finding them by name) but first we should probably do a bit of refactoring so that the situation is less confusing (currently there are a bunch of scheme_kinds for induction principles depending on if the inductive lives in Prop and if target is in prop, for instance `Register False_ind as ind_from_prop for False` is not very nice and we should be able to make `Register False_ind as ind for False` work).

Also fix #18298 

Future work: make `induction` lookup registered schemes instead of hacking names

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Related: 
- https://github.com/coq/coq/issues/10766
- https://github.com/coq/coq/issues/3132
- https://github.com/coq/coq/issues/6794

Also the names for the Scheme command (eg `Scheme Minimality`) being different from the scheme_kind names is not great but needs the scheme_kind unification refactor to make it better.

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 1]
```diff
@@ -1172,7 +1172,21 @@ Exposing constants to OCaml libraries
...
 .. cmd:: Print Registered
-   :undocumented:
+
+   List the currently registered constants.
+
+.. cmd:: Register @qualid__1 as @qualid__2 for @qualid__3
+
+   Make the constant :n:`@qualid__1` accessible to the "scheme"
+   mechanism for scheme kind :n:`@qualid__2` and inductive
+   :n:`@qualid__3`.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst:1182]: Move before `Print Registered` so it reads better (first you register, then you can print).

**Review by jfehrle — COMMENTED** [review]
wording

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 1 above)
Comment (JasonGross) [reply doc/sphinx/proof-engine/vernacular-commands.rst:1182]: I think `Print Registered` is about `Register ... as ...` not `Register ... as ... for...` (for which we have `Print Registered Schemes`) below

Comment (SkySkimmer) [reply doc/sphinx/proof-engine/vernacular-commands.rst:1182]: Yes, schemes are not seen as a subset of regular registrations, they are separate.

**Comment by jdchristensen** [comment]
I tested that with this PR, one can do the following in Basics/Overture.v in the Coq-HoTT library:
```coq
Definition paths_rew {A : Type} {x y : A} (P : A -> Type) (u : P x) (p : x = y)
  : P y
  := transport P p u.

Register Scheme paths_rew as rew for paths.

Definition paths_rew_r {A : Type} {x y : A} (P : A -> Type) (u : P y) (p : x = y)
  : P x
  := transport P p^ u.

Register Scheme paths_rew_r as rew_r for paths.
```
The library builds with no changes, and after unfolding the wrappers, you get `transport`.  Very nice!

I had originally tried this version:
```coq
Register Scheme transport as rew for paths.

Definition transport_V {A : Type} {x y : A} (P : A -> Type) (p : x = y) (u : P y) 
  : P x
  := transport P p^ u.

Register Scheme transport_V as rew_r for paths.
```
The only difference is the order of the arguments `u` and `p`, making them match `transport`.  But then the goals that `rewrite` generates are in a different order, which makes `rewrite 2 lemma` and `rewrite lemma by tactic` not work well.  The core of the HoTT library builds with just one minor change, but the Categories and Classes parts of the library use these variants of rewrite in a number of places, so it's probably best to use the wrappers with the arguments in the other order.

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 1 above)
Comment (SkySkimmer) [reply doc/sphinx/proof-engine/vernacular-commands.rst:1182]: We could merge the printing commands though

Comment (SkySkimmer) [reply doc/sphinx/proof-engine/vernacular-commands.rst:1182]: also I made a mistake it should be `.. cmd:: Register Scheme ...` not just Register

**Comment by SkySkimmer** [comment]
@coqbot run full ci
(pipeline is currently pending, remerge master to be compatible with just now merged overlays)

**Comment by herbelin** [comment]
Is it the question of stabilizing the names of schemes which is blocking?

**Comment by SkySkimmer** [comment]
Pretty much, I really don't want to make this "from_prop" "from_type" stuff user visible.

**Comment by SkySkimmer** [comment]
I worked on this more resulting in c88c10f8d7c713bd421708fdfa33ad8466da0244 (in https://github.com/coq/coq/pull/18549/) which removes the _from_X stuff.
However it still results in weird rules like "rect is expected to be non dependent if the inductive is in Prop, dependent otherwise" so maybe we should stick to the status quo. IDK.

**Comment by SkySkimmer** [comment]
added a topic to https://github.com/coq/coq/wiki/Coq-Call-2024-02-05 about this

**Comment by SkySkimmer** [comment]
I think I managed to handle the elimschemes in a reasonable way, see last commit.

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 2]
```diff
@@ -1168,7 +1168,21 @@ Exposing constants to OCaml libraries
...
+
+.. cmd:: Register Scheme @qualid__1 as @qualid__2 for @qualid__3
+
+   Make the constant :n:`@qualid__1` accessible to the "scheme"
+   mechanism for scheme kind :n:`@qualid__2` and inductive
+   :n:`@qualid__3`.
+
+.. cmd:: Print Registered Schemes
+
+   List the currently registered constants.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst:1182]: This is the same description as for "Print Registered" above.  Could you add something that briefly distinguishes the two commands?  (Are all schemes constants or vice-versa or ...?

Comment (SkySkimmer) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: changed a bit

**Review by ppedrot — APPROVED** [review]
Following offline discussions, I'm not convinced that a string-based mechanism is optimal (remember the curse of globality) but for the time being it should be fine enough.

**Comment by SkySkimmer** [comment]
merge?

**Comment by jfehrle** [comment]
Changelog?

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**